From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lemures Lemniscati <lemures.lemniscati@gmail.com>
Date: Wed, 14 Oct 2020 06:40:46 +0900
Subject: [PATCH] Modify SO_MAJOR in CMakeLists.txt for cygwin


diff --git a/CMakeLists.txt b/CMakeLists.txt
index a1f964da..4f3c57a3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -54,7 +54,7 @@ foreach(line ${configure})
   endforeach()
 endforeach()
 
-math(EXPR SO_MAJOR "${LIBTIFF_CURRENT} - ${LIBTIFF_AGE}")
+math(EXPR SO_MAJOR "${LIBTIFF_CURRENT} - ${LIBTIFF_AGE} + 1")
 set(SO_MINOR "${LIBTIFF_AGE}")
 set(SO_REVISION "${LIBTIFF_REVISION}")
 
