--- origsrc/libtiff-v4.5.0/libtiff/tiffio.h
+++ src/libtiff-v4.5.0/libtiff/tiffio.h
@@ -77,7 +77,7 @@
 typedef tmsize_t tsize_t;   /* i/o size in bytes */
 typedef void *tdata_t;      /* image data ref */
 
-#if !defined(__WIN32__) && (defined(_WIN32) || defined(WIN32))
+#if !defined(__WIN32__) && (defined(_WIN32) || defined(WIN32)) && !defined(__CYGWIN__)
 #define __WIN32__
 #endif
 
@@ -89,7 +89,7 @@
  */
 
 #if defined(_WINDOWS) || defined(__WIN32__) || defined(_Windows)
-#if !defined(__CYGWIN) && !defined(AVOID_WIN32_FILEIO) &&                      \
+#if !defined(__CYGWIN__) && !defined(AVOID_WIN32_FILEIO) &&                      \
     !defined(USE_WIN32_FILEIO)
 #define AVOID_WIN32_FILEIO
 #endif
