--- origsrc/libtiff-v4.6.0/libtiff/mkg3states.c	2023-09-08 18:12:35.000000000 +0900
+++ src/libtiff-v4.6.0/libtiff/mkg3states.c	2024-04-07 22:12:34.511945900 +0900
@@ -229,7 +229,7 @@ int main(int argc, char *argv[])
                 return (-1);
         }
     outputfile = optind < argc ? argv[optind] : "g3states.h";
-    fd = fopen(outputfile, "w");
+    fd = fopen(outputfile, "wb");
     if (fd == NULL)
     {
         fprintf(stderr, "%s: %s: Cannot create output file.\n", argv[0],
--- origsrc/libtiff-v4.6.0/tools/unsupported/tiff2pdf.c	2023-09-08 18:12:35.000000000 +0900
+++ src/libtiff-v4.6.0/tools/unsupported/tiff2pdf.c	2024-04-07 22:13:48.546055800 +0900
@@ -844,6 +844,9 @@ int main(int argc, char **argv)
     {
         outfilename = "-";
         t2p->outputfile = stdout;
+#if defined(HAVE_SETMODE) && defined(O_BINARY)
+        setmode(fileno(stdout), O_BINARY);
+#endif
     }
 
     output = TIFFClientOpen(outfilename, "w", (thandle_t)t2p, t2p_readproc,
--- origsrc/libtiff-v4.6.0/tools/unsupported/tiff2ps.c	2023-09-08 18:12:35.000000000 +0900
+++ src/libtiff-v4.6.0/tools/unsupported/tiff2ps.c	2024-04-07 22:14:19.363865100 +0900
@@ -351,7 +351,7 @@ int main(int argc, char *argv[])
                 }
                 break;
             case 'O': /* XXX too bad -o is already taken */
-                output = fopen(optarg, "w");
+                output = fopen(optarg, "wb");
                 if (output == NULL)
                 {
                     fprintf(stderr, "%s: %s: Cannot open output file.\n",
